src/dfx/src/config/cache.rs

```diff
@@ -286,57 +225,43 @@ pub fn get_binary_path_from_version(version: &str, binary_name: &str) -> DfxResu
     Ok(get_bin_cache(version)?.join(binary_name))
 }
 
-#[context("Failed to get binary '{}' for version '{}'.", name, version)]
-pub fn binary_command_from_version(version: &str, name: &str) -> DfxResult<std::process::Command> {
+pub fn binary_command_from_version(
+    version: &str,
+    name: &str,
+) -> Result<std::process::Command, CacheError> {
     let path = get_binary_path_from_version(version, name)?;
     let cmd = std::process::Command::new(path);
 
     Ok(cmd)
 }
 
-#[context("Failed to list cache versions.")]
-pub fn list_versions() -> DfxResult<Vec<Version>> {
+pub fn list_versions() -> Result<Vec<Version>, CacheError> {
     let root = get_bin_cache_root()?;
     let mut result: Vec<Version> = Vec::new();
 
-    for entry in std::fs::read_dir(&root).with_context(|| {
-        format!(
-            "Failed to read bin cache root content at {}.",
-            root.to_string_lossy()
-        )
-    })? {
-        let entry = entry.with_context(|| {
-            format!(
-                "Failed to read an entry in bin cache root at {}.",
-                root.to_string_lossy()
-            )
-        })?;
+    for entry in dfx_core::fs::read_dir(&root)? {
+        let entry = entry.map_err(CacheError::ReadCacheEntryFailed)?;
         if let Some(version) = entry.file_name().to_str() {
             if version.starts_with('_') {
                 // temp directory for version being installed
                 continue;
             }
-            result.push(
-                Version::parse(version)
-                    .with_context(|| format!("Failed to parse version from {}.", version))?,
-            );
+            result.push(Version::parse(version).map_err(CacheError::MalformedSemverVersion)?);
```

Comment:
Include the problematic version here, as well as the error
```suggestion
            result.push(Version::parse(version).map_err(|e|CacheError::MalformedSemverVersion(version, e))?);
```